cc/dual_net/batching_dual_net.cc

```diff
@@ -0,0 +1,176 @@
+#include "cc/dual_net/batching_dual_net.h"
+
+#include <future>
+#include <queue>
+
+#include "absl/memory/memory.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "cc/check.h"
+
+namespace minigo {
+namespace {
+class BatchingService {
+  struct InferenceData {
+    std::vector<const DualNet::BoardFeatures*> features;
+    std::vector<DualNet::Output*> outputs;
+    std::string* model;
+    absl::Notification* notification;
+  };
+
+ public:
+  explicit BatchingService(std::unique_ptr<DualNet> dual_net)
+      : dual_net_(std::move(dual_net)),
+        num_clients_(0),
+        queue_counter_(0),
+        run_counter_(0),
+        num_runs_(0) {}
+
+  ~BatchingService() {
+    std::cerr << "Ran " << num_runs_ << " batches with an average size of "
+              << static_cast<float>(run_counter_) / num_runs_ << ".\n";
+  }
+
+  void IncrementClientCount() {
+    absl::MutexLock lock(&mutex_);
+    ++num_clients_;
+  }
+
+  void DecrementClientCount() {
+    absl::MutexLock lock(&mutex_);
+    --num_clients_;
+    MaybeRunBatches();
+  }
+
+  void RunMany(std::vector<const DualNet::BoardFeatures*> features,
+                          std::vector<DualNet::Output*> outputs,
+                          std::string* model) {
+    size_t num_features = features.size();
+    MG_CHECK(num_features <= static_cast<size_t>(FLAGS_batch_size));
+    MG_CHECK(num_features == outputs.size());
+
+    absl::Notification notification;
+
+    {
+      absl::MutexLock lock(&mutex_);
+
+      queue_counter_ += num_features;
+      inference_queue_.push({std::move(features), std::move(outputs), model, &notification});
+
+      MaybeRunBatches();
+    }
+
+    notification.WaitForNotification();
+  }
+
+ private:
+  void MaybeRunBatches() EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+    while (size_t batch_size =
+               std::min(queue_counter_ - run_counter_,
+                        static_cast<size_t>(FLAGS_batch_size))) {
+      // Stop if we won't fill a batch yet but more request will come.
```

Comment:
> one of the clients added more than FLAGS_batch_size inferences in one batch which would be a little abnormal.
This is not allowed and checked for in line 49.

But it can happen that RunBatch is called multiple times: the last client has to drain the queue, but his own request might not fit into the same batch anymore. So he will first run a batch for all the other pending clients, and then a small batch just for himself.

Reply:
Reading the code another time I see what's going on.


consider
// Wait for full batch unless all clients have pending requests.
if (static_cast<int>(batch_size) < FLAGS_batch_size && inference_queue_.size() != num_clients_) {
  break;
}

or changing the "but more requests" to "and more requests"

Reply:
Done.